### PR TITLE
Fold final field in ILGen

### PR DESCRIPTION
**Pull request by liqunl**
Folding final fields early can benefit more optimizations.

Folding instance final fields outside of MethodHandle will expose an
optimizer bug that is still under investigation. As for now, fold only
instance fields from MethodHandle.

#4837 

Signed-off-by: Liqun Liu <liqunl@ca.ibm.com>

### CODE REVIEW

**Comment by liqunl** [comment]
@andrewcraik May I ask for a review?

**Comment by andrewcraik** [comment]
Is this safe with OSR? We've generally been trying to move optimizations out of ILGen so why do you want to move this one in?

**Comment by liqunl** [comment]
This is OSR safe. This only folds fields in whitelisted classes. 

Folding primitive fields in ILGen may trigger the folding if `ifcmp`, so some code path can be eliminated earlier and the inliner doesn't have to inline method calls in dead code.

 Folding reference typed fields can help inliner choose better target. For JSR292, we need to fold the next handle in the chain so that we can devirtualize `MethodHandle.invokeExact` to a specific thunk. Currently, this is done in `invariantArgumentPreexistence`, which only runs when `IfNotClassLoadPhaseAndNotProfiling`. This change together with a few upcoming changes will make this optimization run unconditionally.

**Comment by liqunl** [comment]
@andrewcraik Did I answer your question?

File: runtime/compiler/ilgen/Walker.cpp [hunk 1]
```diff
@@ -6302,6 +6312,10 @@ TR_J9ByteCodeIlGenerator::loadStatic(int32_t cpIndex)
          handleSideEffect(treeTopNode);
          genTreeTop(treeTopNode);
          }
+      if (symbol->isFinal() && TR::TransformUtil::canFoldStaticFinalField(comp(), load))
```
Comment (andrewcraik) [inline runtime/compiler/ilgen/Walker.cpp]: `canFoldStaticFinalField` returns yes/no/maybe could you make the comparison explicit for completeness.

**Comment by andrewcraik** [comment]
Question is answered - one thing for code clarity and I think this is good.

File: runtime/compiler/ilgen/Walker.cpp (hunk 1 above)
Comment (liqunl) [reply runtime/compiler/ilgen/Walker.cpp]: Good catch.

Comment (liqunl) [reply runtime/compiler/ilgen/Walker.cpp]: Fixed

**Comment by andrewcraik** [comment]
@liqunl not sure what the deal is with the failed travis build... perhaps a rebase forward would help?

**Comment by liqunl** [comment]
@andrewcraik I'm seeing the same failure in my pbuild, need to take a look, so WIP this PR.

**Comment by liqunl** [comment]
Depends on #6485

**Comment by andrewcraik** [comment]
With #6485 merged is this ready for review?

**Comment by liqunl** [comment]
@andrewcraik There are test failures in my personal build that are sensitive to JIT trace/debug options. I'm still looking at those failures.

**Comment by liqunl** [comment]
@andrewcraik I limit the folding of indirect load to final fields from MethodHandle, it fixes the test failure. Could you take a look again? I'll continue investigation of the test failures with the folding enabled for all whitelist classes.

**Comment by andrewcraik** [comment]
@liqunl do we even want to do it for all whitelisted classes? it is a pretty aggressive thing to do... Limiting to MethodHandles makes sense to me since those are special implementation details.

**Comment by liqunl** [comment]
We can limit to MethodHandles, I was thinking there's a bug lingering somewhere that we need to fix. But we can give that investigation low priority.

**Comment by andrewcraik** [comment]
Not sure if it is a bug or if we were folding a field that isn't really final - worth checking for a non-truly final field in a whitelisted class, but that's what I'd suspect. Can't progress this much more until we know the CI is good.

**Comment by liqunl** [comment]
WIP this PR until the exposed optimizer bug is fixed.

**Comment by andrewcraik** [comment]
@liqunl is this still needed?

**Comment by liqunl** [comment]
@andrewcraik I think yes. Folding final fields early should be good for performance in general. Also, this change does not fold more final fields than we suppose to.

I'll do a rebase and retesting, if everything goes well, maybe we can merge it?

**Comment by liqunl** [comment]
@andrewcraik I rebased the change, it's ready to merge.

**Comment by andrewcraik** [comment]
Change seems reasonable but I think some tracing is warranted and maybe an envar to disable in case of problems.

**Comment by liqunl** [comment]
@andrewcraik Added environment variables to turn off this opt when needed. For traces, the transform utils contain traces.

**Comment by liqunl** [comment]
@andrewcraik Added environment variables to turn off this opt when needed. For traces, the transform utils contain traces.

**Comment by andrewcraik** [comment]
Jenkins test sanity all jdk11

**Comment by liqunl** [comment]
@andrewcraik Sorry I almost forgot this PR. Can we get it merged? It's useful in performance testing in JSR292 when we need a known MethodHandle object before inliner.

File: runtime/compiler/ilgen/Walker.cpp [hunk 2]
```diff
@@ -5195,6 +5195,21 @@ TR_J9ByteCodeIlGenerator::loadInstance(TR::SymbolReference * symRef)
...
+   static char *disableFinalFieldFoldingInILGen = feGetEnv("TR_DisableFinalFieldFoldingInILGen");
+   static char *disableInstanceFinalFieldFoldingInILGen = feGetEnv("TR_DisableInstanceFinalFieldFoldingInILGen");
+   if (!disableFinalFieldFoldingInILGen &&
+       !disableInstanceFinalFieldFoldingInILGen &&
+       address->getOpCode().hasSymbolReference() &&
+       address->getSymbolReference()->hasKnownObjectIndex() &&
+       address->isNonNull())
+      {
+      TR::Node* nodeToRemove = NULL;
+      if (TR::TransformUtil::transformIndirectLoadChain(comp(), dummyLoad, address, address->getSymbolReference()->getKnownObjectIndex(), &nodeToRemove) && nodeToRemove)
```
Comment (andrewcraik) [inline runtime/compiler/ilgen/Walker.cpp:5207]: will this make sure we protect the point with an appropriate set of OSR guards etc in fear analysis?

File: runtime/compiler/ilgen/Walker.cpp [hunk 3]
```diff
@@ -5643,6 +5658,16 @@ TR_J9ByteCodeIlGenerator::loadStatic(int32_t cpIndex)
...
          }
 
+      static char *disableFinalFieldFoldingInILGen = feGetEnv("TR_DisableFinalFieldFoldingInILGen");
+      static char *disableStaticFinalFieldFoldingInILGen = feGetEnv("TR_DisableStaticFinalFieldFoldingInILGen");
+      if (!disableFinalFieldFoldingInILGen &&
+          !disableStaticFinalFieldFoldingInILGen &&
+          symbol->isFinal() &&
+          TR::TransformUtil::canFoldStaticFinalField(comp(), load) == TR_yes)
+         {
+         TR::TransformUtil::foldReliableStaticFinalField(comp(), load);
```
Comment (andrewcraik) [inline runtime/compiler/ilgen/Walker.cpp:5668]: this will just be based on a whitelist of acceptable fields right?

Comment (liqunl) [reply runtime/compiler/ilgen/Walker.cpp:5668]: Yes, unless `-Xjit:unleashStaticFieldFolding` is used or some env var is on to turn on folding blindly, which is usually used in performance experiment

File: runtime/compiler/ilgen/Walker.cpp (hunk 2 above)
Comment (liqunl) [reply runtime/compiler/ilgen/Walker.cpp:5207]: The fields folded here don't need protection. They're fields trusted by VM and changes on them are not expected.

**Comment by liqunl** [comment]
@andrewcraik I've addressed your comments, is this PR good to merge?

**Comment by andrewcraik** [comment]
Good to merge once sanity is done.

**Comment by andrewcraik** [comment]
Jenkins test sanity all jdk15

**Comment by andrewcraik** [comment]
Mac failure is a known issue #9416 

Looks good to go. Merging.

**Pull request merged**